Actually finish the Vulkan Timing Layer.

AWoloszyn opened the pull request:
This layer can be run under a vulkan program, and it
exposes a Perfetto producer (VulkanCPUTiming) that will
export events for each begin/end API call into vulkan.

There will be a follow-up CL that plumbs this information
through into the UI, so that if the layer is available
GAPID can use it to provide this timing information.

pmuetschard commented on core/os/device/remotessh/device.go:
```diff
@@ -314,9 +314,11 @@ func GetConnectedDevice(ctx context.Context, c Configuration) (Device, error) {
 		Auth:            auths,
 		HostKeyCallback: hosts,
 	}
+	log.E(ctx, "Config %+v", sshConfig)
```
Remove, or at least drop the verbosity.

pmuetschard commented on core/os/device/remotessh/device.go:
```diff
@@ -314,9 +314,11 @@ func GetConnectedDevice(ctx context.Context, c Configuration) (Device, error) {
 
 	connection, err := ssh.Dial("tcp", fmt.Sprintf("%s:%d", c.Host, c.Port), sshConfig)
 	if err != nil {
+		log.E(ctx, "Config Err %+v", err)
```
Remove?

pmuetschard commented on core/vulkan/perfetto_producer/perfetto_threadlocal_emitter.inc:
```diff
@@ -0,0 +1,235 @@
+            auto packet = ctx.NewTracePacket();
+            packet->set_timestamp(time);
+            //perfetto::protos::pbzero::InternedData* interned_data = nullptr;
+            //uint64_t nameidd = InternName(packet, *interned_data);
```
remove?

AWoloszyn replied:
Thanks, it went away, then came back. I have been using too many machines

pmuetschard commented on core/vulkan/perfetto_producer/perfetto_threadlocal_emitter.inc:
```diff
@@ -0,0 +1,235 @@
+            //uint64_t nameidd = InternName(packet, *interned_data);
+            packet->set_incremental_state_cleared(true);
+            perfetto::protos::pbzero::ThreadDescriptor* thread_descriptor =
+            packet->set_thread_descriptor();
```
indent

pmuetschard commented on tools/build/third_party/perfetto/perfetto.BUILD:
```diff
@@ -29,7 +29,11 @@ _COPTS_BASE = cc_copts() + [
 
 _COPTS = _COPTS_BASE + [
-    "-DPERFETTO_BUILD_WITH_EMBEDDER",
+    #"-DPERFETTO_BUILD_WITH_EMBEDDER",
```
PERFETTO_BUILD_WITH_EMBEDDER doesn't actually do anything anymore and can be removed.

AWoloszyn replied:
So this has knock-on effects. The build still fails without this code.

pmuetschard commented on tools/build/third_party/perfetto/perfetto.BUILD:
```diff
@@ -156,6 +161,7 @@ cc_library(
         "@gapid//tools/build:windows": [],
         "@gapid//tools/build:darwin": [],
         "@gapid//tools/build:linux": [
+            "src/base/watchdog_posix.cc",
```
remove, this should not be necessary anymore and should only be used by standalone builds.

AWoloszyn replied:
The timing layer links in perfetto to use the client API.  These were required to get everything to build.

pmuetschard commented on tools/build/third_party/perfetto/perfetto.BUILD:
```diff
@@ -218,8 +224,14 @@ cc_library(
-    name = "tracing_core",
+    name = "tracing_ipc",
     srcs = [
+        "src/base/watchdog_posix.cc",
```
remove, this should not be necessary anymore and should only be used by standalone builds.

AWoloszyn replied:
This is needed if we try and link perfetto to the vulkan layer.

pmuetschard commented on tools/build/third_party/perfetto/perfetto.BUILD:
```diff
@@ -218,8 +224,14 @@ cc_library(
 
 cc_library(
-    name = "tracing_core",
+    name = "tracing_ipc",
```
Why did these targets get merged?

AWoloszyn replied:
There are some cross-dependencies between the two targets (once you add the rest of the missing files in) I could find no other way to make all of the builds happy.

pmuetschard commented on gapis/api/templates/vulkan_gfx_api_extras.tmpl:
```diff
@@ -101,6 +101,7 @@ bool Vulkan::replayCreateVkInstanceImpl(Stack* stack,
   std::vector<const char*> layers(
       pCreateInfo->ppEnabledLayerNames,
       pCreateInfo->ppEnabledLayerNames + pCreateInfo->enabledLayerCount);
+  layers.push_back("APITiming");
```
Is this in the right place, shouldn't it be below where we add the VirtualSwapchain layer?

AWoloszyn replied:
This were unintentionally added. I don't want them inserted at all times, will remove.

pmuetschard commented on gapis/api/templates/vulkan_gfx_api_extras.tmpl:
```diff
@@ -101,6 +101,7 @@ bool Vulkan::replayCreateVkInstanceImpl(Stack* stack,
       pCreateInfo->ppEnabledLayerNames,
       pCreateInfo->ppEnabledLayerNames + pCreateInfo->enabledLayerCount);
+  layers.push_back("APITiming");
   layers.reserve(pCreateInfo->enabledLayerCount + 1);
```
Should this now be +2?

pmuetschard commented on gapis/api/vulkan/templates/vulkan_layer.tmpl:
```diff
@@ -477,7 +476,7 @@ VKAPI_ATTR VkResult VKAPI_CALL vkCreateDevice(
   data.get_device_queue = reinterpret_cast<PFN_vkGetDeviceQueue>(get_device_proc_addr(*pDevice, "vkGetDeviceQueue"));
   data.destroy_device = reinterpret_cast<PFN_vkDestroyDevice>(get_device_proc_addr(*pDevice, "vkDestroyDevice"));
-  data.free_command_buffers = reinterpret_cast<PFN_vkFreeCommandBuffers>(get_device_proc_addr(*pDevice, "vkFreeCommandbuffers"));
+  data.free_command_buffers = reinterpret_cast<PFN_vkFreeCommandBuffers>(get_device_proc_addr(*pDevice, "vkFreeCommandBuffers"));
```
Oops. Separate commit?

AWoloszyn replied:
I could, it is related code, as the timing layer is the only thing that actually uses this.

AWoloszyn commented on core/vulkan/perfetto_producer/perfetto_data_source.inc:
```diff
@@ -0,0 +1,114 @@
+ */
+
+#ifndef INCLUDING_DATA_SOURCE_INC__
+#error "Do not include this file directly"
```
This is a fairly standard ".h, .inl" paradigm. You include the .h, it #includes the .inl. It lets us split the public api (.h) with the implementation (.inl) for templated functions.

AWoloszyn commented on core/vulkan/vk_api_timing_layer/cc/layer_helpers.cpp:
```diff
@@ -0,0 +1,48 @@
+#endif
+
+extern "C" {
+    void _layer_dummy_func__() {
```
I should be more explicit (and will add comments) but long story short, to work around the Android loader. This will force this layer to NOT get unloaded once it has been loaded once. (There is a lot of global static state in perfetto that explodes if we ever unload the .so)

AWoloszyn replied:
Added a comment

AWoloszyn commented on core/os/device/deviceinfo/cc/vk.cpp:
```diff
@@ -74,13 +79,14 @@ bool vkLayersAndExtensions(
-        !strcmp(l.layerName, "VirtualSwapchain") ||
-        !strcmp(l.layerName, "ApiTiming")) {
+        !strcmp(l.layerName, "VirtualSwapchain"))
+    {
```
That is correct. It was a mistake to skip it in the first place. I have a follow-up CL that also extends this some more, but I didn't want to over-do it for this one.

AWoloszyn commented on core/vulkan/perfetto_producer/perfetto_threadlocal_emitter.inc:
```diff
@@ -0,0 +1,234 @@
+ */
+
+#ifndef __INCLUDING_PERFETTO_THREADLOCAL_EMITTER_INC__
+#error "Do not include perfetto_threadlocal_emitter.inc directly"
```
The simple answer is "Because it is not a header file". I will expand the comment.

AWoloszyn commented:
> A lot of formatting issues, maybe wrong version of clang-format? see presubmit.
> 
> The clean-up of thread/process abstractions (get_name(), etc) could be a standalone PR?
> 
> Could you add a bit more context in the commit message?

I fixed with the correct clang-format version, and updated to respond to comments.
The change to process/thread is not really a cleanup, but a change, I had to add functionality that was not previously there, in order for this CL to function.
I updated the commit message to be a little bit more descriptive.

AWoloszyn commented on core/vulkan/perfetto_producer/perfetto_data_source.inc:
```diff
@@ -0,0 +1,114 @@
+
+template<typename T>
+void PerfettoProducer<T>::OnSetup(const typename perfetto::DataSourceBase::SetupArgs&) {
+ // Todo: Anything needed here?
```
Removed the TODO

hevrard commented on core/vulkan/vk_api_timing_layer/cc/layer_helpers.cpp:
```diff
@@ -0,0 +1,53 @@
+// _layer_dummy_func__ is marked __attribute__((constructor))
+// this means on .so open, it will be called. Once that happens,
+// we create a dummy struct, which on Android and Linux,
+// Forces the layer to never be unloaded. There is some global
```
Typo "Forces" -> "forces"
